crates/biome_js_analyze/src/lint/correctness/no_undeclared_dependencies.rs

```diff
@@ -34,32 +40,208 @@ declare_lint_rule! {
     /// ```js,ignore
     /// import assert from "node:assert";
     /// ```
+    ///
+    /// ## Options
+    ///
+    /// This rule supports the following options:
+    /// - `devDependencies`: If set to `false`, then the rule will show an error when `devDependencies` are imported. Defaults to `true`.
+    /// - `peerDependencies`: If set to `false`, then the rule will show an error when `peerDependencies` are imported. Defaults to `true`.
+    /// - `optionalDependencies`: If set to `false`, then the rule will show an error when `optionalDependencies` are imported. Defaults to `true`.
+    ///
+    /// You can set the options like this:
+    /// ```json
+    /// {
+    ///   "options": {
+    ///     "devDependencies": false,
+    ///     "peerDependencies": false,
+    ///     "optionalDependencies": false
+    ///   }
+    /// }
+    /// ```
+    ///
+    /// You can also use an array of globs instead of literal booleans:
+    /// ```json
+    /// {
+    ///   "options": {
+    ///     "devDependencies": ["**/*.test.js", "**/*.spec.js"]
+    ///   }
+    /// }
+    /// ```
+    ///
+    /// When using an array of globs, the setting will be set to `true` (no errors reported)
+    /// if the name of the file being linted (i.e. not the imported file/module) matches a single glob
+    /// in the array, and `false` otherwise.
     pub NoUndeclaredDependencies {
         version: "1.6.0",
         name: "noUndeclaredDependencies",
         language: "js",
+        sources: &[
+            RuleSource::EslintImport("no-extraneous-dependencies"),
+        ],
         recommended: false,
     }
 }
 
+#[derive(Clone, Debug, serde::Deserialize, serde::Serialize)]
+#[serde(untagged)]
+enum DependencyAvailability {
+    /// Dependencies are always available or unavailable.
+    Bool(bool),
+
+    /// Dependencies are available in files that matches any of the globs.
+    Patterns(Box<[RestrictedGlob]>),
+}
+
+impl Default for DependencyAvailability {
+    fn default() -> Self {
+        Self::Bool(true)
+    }
+}
+
+impl PartialEq for DependencyAvailability {
+    fn eq(&self, other: &Self) -> bool {
+        match self {
+            Self::Bool(a) => match other {
+                Self::Bool(b) => a == b,
+                _ => false,
+            },
+            Self::Patterns(a) => match other {
+                Self::Patterns(b) => {
+                    a.len() == b.len()
+                        && a.iter()
+                            .zip(b.iter())
+                            .all(|(a, b)| a.to_string() == b.to_string())
+                }
+                _ => false,
+            },
+        }
+    }
+}
```

Comment:
You are the first consumer of `RestrictedGlob` and thanks to you I discovered that `RestrictedGlob` don't implement `Eq` and `PartialEq`. This sill be fixed by #4388. Once merged you will be able to derive `Eq`, `PartialEq`, and `Hash`.

Reply:
Just curious, should we derive Hash on rule options struct even if it's not required? (I noticed Eq and PartialEq are required, but Hash is not)

Reply:
Merged :)

Reply:
Merged main and replaced the impls with derive in a8c87c4.